src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

```diff
@@ -181,23 +193,49 @@ private async Task DoReceive()
                     // Pipe consumer is shut down, do we stop writing
                     break;
                 }
-            }
-        }
-        catch (SocketException ex) when (IsConnectionResetError(ex.SocketErrorCode))
-        {
-            // This could be ignored if _shutdownReason is already set.
-            error = new ConnectionResetException(ex.Message, ex);
 
-            // There's still a small chance that both DoReceive() and DoSend() can log the same connection reset.
-            // Both logs will have the same ConnectionId. I don't think it's worthwhile to lock just to avoid this.
-            if (!_socketDisposed)
-            {
-                SocketsLog.ConnectionReset(_logger, this);
+                bool IsNormalCompletion(SocketOperationResult result)
+                {
+                    if (!result.HasError)
+                    {
+                        return true;
+                    }
+
+                    if (IsConnectionResetError(result.SocketError!.SocketErrorCode))
+                    {
+                        // This could be ignored if _shutdownReason is already set.
+                        var ex = result.SocketError!;
+                        error = new ConnectionResetException(ex.Message, ex);
+
+                        // There's still a small chance that both DoReceive() and DoSend() can log the same connection reset.
+                        // Both logs will have the same ConnectionId. I don't think it's worthwhile to lock just to avoid this.
+                        if (!_socketDisposed)
+                        {
+                            SocketsLog.ConnectionReset(_logger, this);
+                        }
+
+                        return false;
+                    }
+
+                    if (IsConnectionAbortError(result.SocketError!.SocketErrorCode))
+                    {
+                        // This exception should always be ignored because _shutdownReason should be set.
+                        error = result.SocketError!;
+
+                        if (!_socketDisposed)
+                        {
+                            // This is unexpected if the socket hasn't been disposed yet.
+                            SocketsLog.ConnectionError(_logger, this, error);
+                        }
+
+                        return false;
+                    }
+
+                    return true;
```

Comment:
This should never return true without setting `error` to some non-null value.

Reply:
I misread this initially. If `result.HasError`, we should always be returning false. And this should never return *false* without setting error to some non-null value.

Reply:
We are now ignoring "unknown" errors. There are no tests for this, because these errors are unknown and therefor we don't know how to reproduce it, but we definitely should not be swallowing these.

Assuming `result.BytesTransferred` is 0 given an unkown error (which seems like a safe bet), then the error will get logged as a FIN which is terrible for debugging.

```suggestion
                    // This is unexpected.
                    error = result.SocketError;
                    SocketsLog.ConnectionError(_logger, this, error);
```